dali/kernels/slice/slice_gpu.h

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SLICE_SLICE_GPU_H_
+#define DALI_KERNELS_SLICE_SLICE_GPU_H_
+
+#include <cuda_runtime.h>
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/convert.h"
+#include "dali/core/dev_array.h"
+#include "dali/core/error_handling.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+
+template <std::size_t Dims>
+struct SliceArgsDev {
+  DeviceArray<int64_t, Dims> in_strides;
+  DeviceArray<int64_t, Dims> out_strides;
+  DeviceArray<int64_t, Dims> anchor;
+};
+
+template <typename OutputType, typename InputType, std::size_t Dims>
+__global__ void SliceKernel(OutputType *__restrict__ output,
+                            const InputType *__restrict__ input,
+                            SliceArgsDev<Dims> slice_args_dev,
+                            unsigned int total_pixels) {
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= total_pixels) {
+    return;
+  }
+
+  unsigned int out_idx = idx;
+  unsigned int in_idx = 0;
+  for (std::size_t d = 0; d < Dims; d++) {
+    unsigned int i_d = idx / slice_args_dev.out_strides[d];
+    idx = idx % slice_args_dev.out_strides[d];
+    in_idx += (slice_args_dev.anchor[d] + i_d) * slice_args_dev.in_strides[d];
+  }
+  output[out_idx] = clamp<OutputType>(input[in_idx]);
+}
+
+template <std::size_t Dims>
+struct SliceArgs {
+  std::array<int64_t, Dims> anchor;
+  std::array<int64_t, Dims> shape;
+};
+
+template <std::size_t Dims, typename Shape>
+std::array<int64_t, Dims> GetStrides(const Shape& shape) {
+  std::array<int64_t, Dims> strides;
+  strides[Dims - 1] = 1;
+  for (std::size_t d = Dims - 1; d > 0; d--) {
+    strides[d - 1] = strides[d] * shape[d];
+  }
+  return strides;
+}
+
+template <typename InputType, typename OutputType, std::size_t Dims>
+class DLL_PUBLIC SliceGPU {
+ public:
+  SliceGPU() = default;
+
+  DLL_PUBLIC KernelRequirements Setup(KernelContext &context,
+                                      const InListGPU<InputType, Dims> &in,
+                                      const std::vector<SliceArgs<Dims>> &slice_args) {
+    KernelRequirements req;
+
+    DALI_ENFORCE(slice_args.size() == static_cast<std::size_t>(in.size()),
+      "Number of samples and size of slice arguments should match");
+
+    std::vector<TensorShape<Dims>> output_shapes;
```

Comment:
It'd be better to create a `TensorListShape<Dims>` here and resize it to accommodate `in.size()` samples.

Reply:
done